File: Hosts/Java/ToPython3/extended/Main_test.java
package extended;

import metaffi.MetaFFIHandle;
import metaffi_host.*;

public class Main_test
{
	public static void main(String[] args)
	{
		try
		{
 			metaffi_host.collections.load("collections_MetaFFIGuest");
 			metaffi_host.extended_testModule.load("extended_test_MetaFFIGuest");
            metaffi_host.metaffi_objects.load("metaffi_objects_MetaFFIGuest");
            System.out.println("--- test_property");
            test_property();

            System.out.println("--- test_positional_or_named");
			test_positional_or_named();

            System.out.println("--- test_positional_or_named_multi_type_hint");
			test_positional_or_named_multi_type_hint();

            System.out.println("--- test_list_args");
			test_list_args();

            System.out.println("--- test_dict_args");
            test_dict_args();

            System.out.println("--- test_named_only");
			test_named_only();

            System.out.println("--- test_positional_only");
			test_positional_only();

            System.out.println("--- test_arg_positional_arg_named");
			test_arg_positional_arg_named();
		}
		catch(Exception e)
		{
			System.out.println("Unexpected exception has occurred during tests");
			e.printStackTrace();
			System.exit(1);
		}
	}

    private static void test_property() throws Exception
    {
        var ext = new extended_test();

        ext.Setx(4);
        var x = ext.Getx();

        if(x != 4)
        {
            throw new Exception(String.format("Expected to return 4, returned %d", x));
        }
    }

	private static void test_positional_or_named() throws Exception
	{
        var ext = new extended_test();

        var str = ext.positional_or_named("PositionalOrNamed"); // positional_or_named('PositionalOrNamed')
        if(!str.equals("PositionalOrNamed")){
            throw new Exception(String.format("\"str\" != %s", str));
        }
	}

	private static void test_positional_or_named_multi_type_hint() throws Exception
	{
		var ext = new extended_test();

        var str = ext.positional_or_named_multi_type_hint("arg1"); // positional_or_named_multi_type_hint('arg1')
        if(!str.equals("arg1")){
            throw new Exception(String.format("\"arg1\" != %s", str));
        }

        str = ext.positional_or_named_multi_type_hint("arg1", "arg2"); // positional_or_named_multi_type_hint('arg1')
        if(!str.equals("arg1 arg2")){
            throw new Exception(String.format("\"arg1 arg2\" != %s", str));
        }
	}

	private static void test_list_args() throws Exception
	{
		var ext = new extended_test();

        MetaFFIHandle listHandle = (MetaFFIHandle)ext.list_args(); // list_args()

        var lst = new metaffi_host.UserList(listHandle);

        var item0 = lst.__getitem__(0); // get first item
        if (!((String)item0).equals("default")){
            throw new Exception(String.format("\"default\" != %s", (String)item0));
        }

		//---------------

        listHandle = (MetaFFIHandle)ext.list_args("None Default"); // list_args()

        lst = new metaffi_host.UserList(listHandle);

        item0 = lst.__getitem__(0); // get first item
        if (!((String)item0).equals("None Default")){
            throw new Exception(String.format("\"None Default\" != %s", (String)item0));
        }

        //---------------

		var list_args = new metaffi_host.metaffi_positional_args();
		list_args.set_arg("arg1");
		list_args.set_arg("arg2");
		list_args.set_arg("arg3");

        listHandle = (MetaFFIHandle)ext.list_args("None-Default 2", list_args.getHandle()); // list_args()

        lst = new metaffi_host.UserList(listHandle);

        item0 = lst.__getitem__(0); // get first item
        if (!((String)item0).equals("None-Default 2")){
            throw new Exception(String.format("\"None-Default 2\" != %s", (String)item0));
        }

        var item1 = lst.__getitem__(1);
        if (!((String)item1).equals("arg1")){
            throw new Exception(String.format("\"arg1\" != %s", (String)item1));
        }

        var item2 = lst.__getitem__(2);
        if (!((String)item2).equals("arg2")){
            throw new Exception(String.format("\"arg2\" != %s", (String)item2));
        }

        var item3 = lst.__getitem__(3);
        if (!((String)item3).equals("arg3")){
            throw new Exception(String.format("\"arg3\" != %s", (String)item3));
        }
	}

	private static void test_dict_args() throws Exception
    {
        var ext = new extended_test();

        MetaFFIHandle lstHandle = (MetaFFIHandle)ext.dict_args();
        var lst = new metaffi_host.UserList(lstHandle);

        var item0 = lst.__getitem__(0);
        if (!((String)item0).equals("default")){
            throw new Exception(String.format("\"default\" != %s", (String)item0));
        }

        //-------

		lstHandle = (MetaFFIHandle)ext.dict_args("none-default");
		lst = new metaffi_host.UserList(lstHandle);
		item0 = lst.__getitem__(0);
        if (!((String)item0).equals("none-default")){
            throw new Exception(String.format("\"none-default\" != %s", (String)item0));
        }

        //-------

        var keyword_args = new metaffi_keyword_args();
        keyword_args.set_arg("key1", "val1");

        lstHandle = (MetaFFIHandle)ext.dict_args("none-default", keyword_args.getHandle());

		lst = new metaffi_host.UserList(lstHandle);
        item0 = lst.__getitem__(0);
        if (!((String)item0).equals("none-default")){
            throw new Exception(String.format("\"none-default\" != %s", (String)item0));
        }

        lst = new metaffi_host.UserList(lstHandle);
        var item1 = lst.__getitem__(1);
        if (!((String)item1).equals("key1")){
            throw new Exception(String.format("\"key1\" != %s", (String)item1));
        }

        lst = new metaffi_host.UserList(lstHandle);
        var item2 = lst.__getitem__(2);
        if (!((String)item2).equals("val1")){
            throw new Exception(String.format("\"val1\" != %s", (String)item2));
        }
    }

	private static void test_named_only() throws Exception
    {
        var ext = new extended_test();
        var keyword_args = new metaffi_keyword_args();
        keyword_args.set_arg("named", "test");

        var res = ext.named_only(keyword_args.getHandle());

		if (!res.equals("test")){
            throw new Exception(String.format("\"test\" != %s", res));
        }
    }

    private static void test_positional_only() throws Exception
    {
        var ext = new extended_test();

        String res = ext.positional_only("word1");

        if (!res.equals("word1 default")){
            throw new Exception(String.format("\"word1 default\" != %s", res));
        }

        res = ext.positional_only("word1", "word2");

        if (!res.equals("word1 word2")){
            throw new Exception(String.format("\"word1 word2\" != %s", res));
        }
    }

    private static void test_arg_positional_arg_named() throws Exception
    {
        var ext = new extended_test();
		var lstHandle = ext.arg_positional_arg_named();

		var lst = new metaffi_host.UserList(lstHandle);
		var item0 = lst.__getitem__(0);
        if (!((String)item0).equals("default")){
            throw new Exception(String.format("\"default\" != %s", (String)item0));
        }

		//-----

		lstHandle = (MetaFFIHandle)ext.arg_positional_arg_named("positional arg");
        lst = new metaffi_host.UserList(lstHandle);
        item0 = lst.__getitem__(0);
        if (!((String)item0).equals("positional arg")){
            throw new Exception(String.format("\"default\" != %s", (String)item0));
        }

        //-----

	    var listArgs = new metaffi_positional_args();
	    listArgs.set_arg("var positional arg");

        lstHandle = (MetaFFIHandle)ext.arg_positional_arg_named("positional arg", listArgs.getHandle());

        lst = new metaffi_host.UserList(lstHandle);
        item0 = lst.__getitem__(0);
        if (!((String)item0).equals("positional arg")){
            throw new Exception(String.format("\"default\" != %s", (String)item0));
        }

        var item1 = lst.__getitem__(1);
        if (!((String)item1).equals("var positional arg")){
            throw new Exception(String.format("\"var positional arg\" != %s", (String)item1));
        }

        //-----
        var keyword_args = new metaffi_keyword_args();
        keyword_args.set_arg("key1", "val1");

        lstHandle = ext.arg_positional_arg_named("positional arg", listArgs.getHandle(), keyword_args.getHandle());

        lst = new metaffi_host.UserList(lstHandle);
        item0 = lst.__getitem__(0);
        if (!((String)item0).equals("positional arg")){
            throw new Exception(String.format("\"default\" != %s", (String)item0));
        }

        item1 = lst.__getitem__(1);
        if (!((String)item1).equals("var positional arg")){
            throw new Exception(String.format("\"var positional arg\" != %s", (String)item1));
        }

        var item2 = lst.__getitem__(2);
        if (!((String)item2).equals("key1")){
            throw new Exception(String.format("\"key1\" != %s", (String)item2));
        }

        var item3 = lst.__getitem__(3);
        if (!((String)item3).equals("val1")){
            throw new Exception(String.format("\"val1\" != %s", (String)item3));
        }
    }
}

File: Hosts/Go/ToPython3/extended/Main_test.go
package sanity

import (
	"test/extended_test"
	"test/collections"
	"test/metaffi_objects"
	"os"
	"testing"
	. "github.com/MetaFFI/lang-plugin-go/go-runtime"
	"fmt"
)

// --------------------------------------------------------------------
func TestMain(m *testing.M) {
	extended_test.MetaFFILoad("extended_test_MetaFFIGuest")
	collections.MetaFFILoad("collections_MetaFFIGuest")
	metaffi_objects.MetaFFILoad("metaffi_objects_MetaFFIGuest")

	exitVal := m.Run()

	os.Exit(exitVal)
}
//--------------------------------------------------------------------
func TestPositionalOrNamed(t *testing.T) {
	ext, err := extended_test.NewExtended_Test()
	if err != nil{ t.Fatal(err) }

	err = ext.Setx_MetaFFISetter(4)
	if err != nil{ t.Fatal(err) }

	x, err := ext.Getx_MetaFFIGetter()
	if err != nil{ t.Fatal(err) }

	if x != 4{
		t.Fatalf("x != 4. x == %v", x)
	}

	// PositionalOrNamed()

	str, err := ext.Positional_Or_Named("PositionalOrNamed") // positional_or_named('PositionalOrNamed')
	if err != nil{ t.Fatal(err) }

	if str != "PositionalOrNamed"{
		t.Fatalf("\"str\" != %v", str)
	}
}
//--------------------------------------------------------------------
func TestPositionalOrNamedMultiTypeHint1(t *testing.T){

	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

	str, err := ext.Positional_Or_Named_Multi_Type_Hint1("arg1") // positional_or_named_multi_type_hint('arg1')
    if err != nil{ t.Fatal(err) }
    if str != "arg1"{
        t.Fatalf("\"arg1\" != %v", str)
    }
}
//--------------------------------------------------------------------
func TestPositionalOrNamedMultiTypeHint(t *testing.T){

	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

	str, err := ext.Positional_Or_Named_Multi_Type_Hint("arg1", "arg2")  // positional_or_named_multi_type_hint('arg1', 'arg2')
    if err != nil{ t.Fatal(err) }
    if str != "arg1 arg2"{
        t.Fatalf("\"arg1 arg2\" != %v", str)
    }
}
//--------------------------------------------------------------------
func TestListArgs1(t *testing.T){

	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

	listHandle, err := ext.List_Args1() // list_args()
    if err != nil{ t.Fatal(err) }

    lst := collections.UserList{}
    lst.SetHandle(listHandle.(Handle))

    item0, err := lst.U___Getitem____(0) // get first item
    if err != nil{ t.Fatal(err) }

    if item0.(string) != "default"{
        t.Fatalf("\"default\" != %v", item0)
    }
}
//--------------------------------------------------------------------
func TestListArgs2(t *testing.T){

	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

	listHandle, err := ext.List_Args2("None Default") // list_args('None Default')
    if err != nil{ t.Fatal(err) }

    lst := collections.UserList{}
    lst.SetHandle(listHandle.(Handle))

    item0, err := lst.U___Getitem____(0) // get first item
    if err != nil{ t.Fatal(err) }

    if item0.(string) != "None Default"{
        t.Fatalf("\"None Default\" != %v", item0)
    }
}
//--------------------------------------------------------------------
func TestListArgs(t *testing.T){

	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

    listArgs, err := metaffi_objects.NewMetaffi_Positional_Args()
    if err != nil{ t.Fatal(err) }

    err = listArgs.Set_Arg("arg1")
    if err != nil{ t.Fatal(err) }

    err = listArgs.Set_Arg("arg2")
    if err != nil{ t.Fatal(err) }

    err = listArgs.Set_Arg("arg3")
    if err != nil{ t.Fatal(err) }

    listHandle, err := ext.List_Args("None-Default 2", listArgs.GetHandle())
    if err != nil{ t.Fatal(err) }

    lst := collections.UserList{}
    lst.SetHandle(listHandle.(Handle))

	len, err := lst.U___Len____()
	if err != nil{ t.Fatal(err) }
	fmt.Printf("returned list size: %v\n", len)

    item0, err := lst.U___Getitem____(0) // get first item
    if err != nil{ t.Fatal(err) }
    if item0.(string) != "None-Default 2"{
        t.Fatalf("\"None-Default 2\" != %v", item0)
    }

    item1, err := lst.U___Getitem____(1) // get first item
    if err != nil{ t.Fatal(err) }
    if item1.(string) != "arg1"{
        t.Fatalf("\"arg1\" != %v", item1)
    }

    item2, err := lst.U___Getitem____(2) // get first item
    if err != nil{ t.Fatal(err) }
    if item2.(string) != "arg2"{
        t.Fatalf("\"arg2\" != %v", item2)
    }

    item3, err := lst.U___Getitem____(3) // get first item
    if err != nil{ t.Fatal(err) }
    if item3.(string) != "arg3"{
        t.Fatalf("\"arg3\" != %v", item3)
    }

}
//--------------------------------------------------------------------
func TestDictArgs1(t *testing.T){

	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

	res, err := ext.Dict_Args1()
	if err != nil{ t.Fatal(err) }

	lst := collections.UserList{}
    lst.SetHandle(res.(Handle))

    item0, err := lst.U___Getitem____(0) // get first item
    if err != nil{ t.Fatal(err) }
    if item0.(string) != "default"{
        t.Fatalf("\"default\" != %v", item0)
    }
}
//--------------------------------------------------------------------
func TestDictArgs2(t *testing.T){

	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

	res, err := ext.Dict_Args2("none-default")
	if err != nil{ t.Fatal(err) }

	lst := collections.UserList{}
    lst.SetHandle(res.(Handle))

    item0, err := lst.U___Getitem____(0) // get first item
    if err != nil{ t.Fatal(err) }
    if item0.(string) != "none-default"{
        t.Fatalf("\"none-default\" != %v", item0)
    }
}
//--------------------------------------------------------------------
func TestDictArgs(t *testing.T){

	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

	dictArgs, err := metaffi_objects.NewMetaffi_Keyword_Args()
	if err != nil{ t.Fatal(err) }

	err = dictArgs.Set_Arg("key1", "val1");
	if err != nil{ t.Fatal(err) }

	res, err := ext.Dict_Args("none-default", dictArgs.GetHandle())
	if err != nil{ t.Fatal(err) }

	lst := collections.UserList{}
    lst.SetHandle(res.(Handle))

    item0, err := lst.U___Getitem____(0) // get first item
    if err != nil{ t.Fatal(err) }
    if item0.(string) != "none-default"{
        t.Fatalf("\"none-default\" != %v", item0)
    }

    item1, err := lst.U___Getitem____(1) // get first item
    if err != nil{ t.Fatal(err) }
    if item1.(string) != "key1"{
        t.Fatalf("\"key1\" != %v", item1)
    }

    item2, err := lst.U___Getitem____(2) // get first item
    if err != nil{ t.Fatal(err) }
    if item2.(string) != "val1"{
        t.Fatalf("\"val1\" != %v", item2)
    }
}
//--------------------------------------------------------------------
func TestNamedOnly(t *testing.T){

	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

    kwargs, err := metaffi_objects.NewMetaffi_Keyword_Args()
    if err != nil{ t.Fatal(err) }
    kwargs.Set_Arg("named", "test")

	res, err := ext.Named_Only(kwargs.GetHandle())
	if err != nil{ t.Fatal(err) }

	if res != "test"{
		t.Fatalf("\"test\" != %v", res)
	}
}
//--------------------------------------------------------------------
func TestPositionalOnly1(t *testing.T){
	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

    res, err := ext.Positional_Only1("word1")
    if err != nil{ t.Fatal(err) }

    if res != "word1 default"{
        t.Fatalf("\"word1 default\" != %v", res)
    }
}
//--------------------------------------------------------------------
func TestPositionalOnly(t *testing.T){
	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

    res, err := ext.Positional_Only("word1", "word2")
    if err != nil{ t.Fatal(err) }

    if res != "word1 word2"{
        t.Fatalf("\"word1 word2\" != %v", res)
    }
}
//--------------------------------------------------------------------
func TestArgPositionalArgNamed1(t *testing.T){
	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

    lstHandle, err := ext.Arg_Positional_Arg_Named1()
    if err != nil{ t.Fatal(err) }

    lst := collections.UserList{}
    lst.SetHandle(lstHandle.(Handle))

    item0, err := lst.U___Getitem____(0) // get first item
    if err != nil{ t.Fatal(err) }
    if item0.(string) != "default"{
        t.Fatalf("\"default\" != %v", item0)
    }

}
//--------------------------------------------------------------------
func TestArgPositionalArgNamed2(t *testing.T){
	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

    lstHandle, err := ext.Arg_Positional_Arg_Named2("positional arg")
    if err != nil{ t.Fatal(err) }

    lst := collections.UserList{}
    lst.SetHandle(lstHandle.(Handle))

    item0, err := lst.U___Getitem____(0) // get first item
    if err != nil{ t.Fatal(err) }
    if item0.(string) != "positional arg"{
        t.Fatalf("\"positional arg\" != %v", item0)
    }

}
//--------------------------------------------------------------------
func TestArgPositionalArgNamed3(t *testing.T){
	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

    listArgs, err := metaffi_objects.NewMetaffi_Positional_Args()
    if err != nil{ t.Fatal(err) }

    err = listArgs.Set_Arg("var positional arg")
    if err != nil{ t.Fatal(err) }

    lstHandle, err := ext.Arg_Positional_Arg_Named3("positional arg", listArgs.GetHandle())
    if err != nil{ t.Fatal(err) }

    lst := collections.UserList{}
    lst.SetHandle(lstHandle.(Handle))

    item0, err := lst.U___Getitem____(0) // get first item
    if err != nil{ t.Fatal(err) }
    if item0.(string) != "positional arg"{
        t.Fatalf("\"positional arg\" != %v", item0)
    }

    item1, err := lst.U___Getitem____(1) // get first item
    if err != nil{ t.Fatal(err) }
    if item1.(string) != "var positional arg"{
        t.Fatalf("\"var positional arg\" != %v", item1)
    }
}
//--------------------------------------------------------------------
func TestArgPositionalArgNamed(t *testing.T){
	ext, err := extended_test.NewExtended_Test()
    if err != nil{ t.Fatal(err) }

    listArgs, err := metaffi_objects.NewMetaffi_Positional_Args()
    if err != nil{ t.Fatal(err) }

    err = listArgs.Set_Arg("var positional arg")
    if err != nil{ t.Fatal(err) }

    kwargs, err := metaffi_objects.NewMetaffi_Keyword_Args()
	if err != nil{ t.Fatal(err) }

	err = kwargs.Set_Arg("key1", "val1")
	if err != nil{ t.Fatal(err) }

    lstHandle, err := ext.Arg_Positional_Arg_Named("positional arg", listArgs.GetHandle(), kwargs.GetHandle())
    if err != nil{ t.Fatal(err) }

    lst := collections.UserList{}
    lst.SetHandle(lstHandle.(Handle))

    item0, err := lst.U___Getitem____(0) // get first item
    if err != nil{ t.Fatal(err) }
    if item0.(string) != "positional arg"{
        t.Fatalf("\"positional arg\" != %v", item0)
    }

    item1, err := lst.U___Getitem____(1) // get first item
    if err != nil{ t.Fatal(err) }
    if item1.(string) != "var positional arg"{
        t.Fatalf("\"var positional arg\" != %v", item1)
    }

    item2, err := lst.U___Getitem____(2) // get first item
    if err != nil{ t.Fatal(err) }
    if item2.(string) != "key1"{
        t.Fatalf("\"key1\" != %v", item2)
    }

    item3, err := lst.U___Getitem____(3) // get first item
    if err != nil{ t.Fatal(err) }
    if item3.(string) != "val1"{
        t.Fatalf("\"val1\" != %v", item3)
    }
}
//--------------------------------------------------------------------